### PR TITLE
Reject process instance migration if element has an incident

### LINKED ISSUE #15440: Reject process instance migration if element has an incident
In the `ProcessInstanceMigration` command processor, reject the command if:
- an element is active that has an incident raised on it

We can't yet migrate incidents, so this could lead to undefined behavior.

Rejecting a command includes both appending a rejection with the rejection writer, as well as responding the rejection with the response writer.

Some tests must showcase this behavior.

Blocked by:
- https://github.com/camunda/zeebe/issues/15111

### PR DESCRIPTION
**Pull request by berkaycanbc**
## Related issues

closes #15440 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/MigrateProcessInstanceRejectionTest.java [hunk 1]
```diff
@@ -146,6 +154,168 @@ public void shouldRejectCommandWhenActiveElementIsNotMapped() {
...
+        .addMappingInstruction("A", "A")
+        .migrate();
+
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_MIGRATED)
+                .withProcessInstanceKey(processInstanceKey)
+                .withElementType(BpmnElementType.PROCESS)
+                .findAny())
+        .describedAs("Expected to have migrated the process instance")
+        .isPresent();
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/MigrateProcessInstanceRejectionTest.java:235]: 👍 Nice additional assertion

File: engine/src/main/java/io/camunda/zeebe/engine/processing/BpmnProcessors.java [hunk 2]
```diff
@@ -236,7 +236,8 @@ private static void addProcessInstanceMigrationStreamProcessors(
             processingState.getElementInstanceState(),
             processingState.getProcessState(),
             processingState.getJobState(),
-            processingState.getVariableState()));
+            processingState.getVariableState(),
+            processingState.getIncidentState()));
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/BpmnProcessors.java:240]: This is growing a bit out of proportion 😆 

🔧 We could simply pass `processingState` to the processor, allowing the constructor to pick the needed parts.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java [hunk 3]
```diff
@@ -177,6 +188,16 @@ private void tryMigrateElementInstance(
           elementInstanceRecord.getProcessInstanceKey(), elementInstanceRecord.getElementId());
     }
 
+    final boolean hasProcessIncident =
+        incidentState.getProcessInstanceIncidentKey(elementInstance.getKey()) != MISSING_INCIDENT;
+    final boolean hasJobIncident =
+        incidentState.getJobIncidentKey(elementInstance.getJobKey()) != MISSING_INCIDENT;
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:194]: 🔧 This lookup is only necessary if `elementInstance.getJobKey() > -1`

**Review by korthout — APPROVED** [review]
Nice work @berkaycanbc 🔥 

LGTM 👍

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java (hunk 3 above)
Comment (berkaycanbc) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:194]: @korthout I thought about the same thing but it looked a bit ugly in an if statement or when you do a separate if checks. We can refactor it something like:

```java
final boolean hasIncident =
        incidentState.getProcessInstanceIncidentKey(elementInstance.getKey()) != MISSING_INCIDENT ||
        incidentState.getJobIncidentKey(elementInstance.getJobKey()) != MISSING_INCIDENT;
```
WDYT?

Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:194]: 💭 Performance > Looks

```java
final boolean hasIncident =
        incidentState.getProcessInstanceIncidentKey(elementInstance.getKey()) != MISSING_INCIDENT ||
        (elementInstance.getJobKey() > -1L && incidentState.getJobIncidentKey(elementInstance.getJobKey()) != MISSING_INCIDENT;)
```

Comment (berkaycanbc) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:194]: Ahh, that is even better! 🚀

File: engine/src/main/java/io/camunda/zeebe/engine/processing/BpmnProcessors.java (hunk 2 above)
Comment (berkaycanbc) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/BpmnProcessors.java:240]: Makes total sense. 👍 Updated!

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/MigrateProcessInstanceRejectionTest.java [hunk 4]
```diff
@@ -146,6 +154,168 @@ public void shouldRejectCommandWhenActiveElementIsNotMapped() {
...
+
+    assertThat(rejectionRecord)
+        .hasIntent(ProcessInstanceMigrationIntent.MIGRATE)
+        .hasRejectionType(RejectionType.INVALID_STATE)
+        .hasRejectionReason(
+            String.format(
+                """
+                Expected to migrate process instance '%d' \
+                but active element with id 'A' has an incident. \
+                Elements cannot be migrated with an incident. \
```
[the remark this answers is not in the document]
Comment (berkaycanbc) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/MigrateProcessInstanceRejectionTest.java:214]: Updated!

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java (hunk 3 above)
Comment (berkaycanbc) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:194]: Updated!

**Comment by berkaycanbc** [comment]
bors merge

**Pull request merged**